### PR TITLE
add plan area changes table and populate them from plan tables field

### PR DESCRIPTION
**Pull request by muamichali**
starts issue #403, just need to add some tests

### CODE REVIEW

File: server/bin/populate_plan_area_changes [hunk 1]
```diff
@@ -0,0 +1,30 @@
+
```
Comment (TomerMe2) [inline server/bin/populate_plan_area_changes:1]: Please rename this file to have a file format of js (It currently has no file format)

File: server/migrations/20210722090328_create_plan_area_changes_table.js [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+        t.text('approved_state').notNullable();
+        t.text('change_to_approved_state').notNullable();
+        t.text('total_in_detailed_plan').notNullable();
+        t.text('total_in_mitaarit_plan').notNullable();
+        t.text('remarks').notNullable();
+    });
+};
+
+exports.down = async function(knex) {
+    await knex.schema.dropTableIfExists('file');
```
Comment (TomerMe2) [inline server/migrations/20210722090328_create_plan_area_changes_table.js:16]: change 'file' into 'plan_area_changes'

File: server/tests/unit/model/plan_area_changes.test.js [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+		instance = new PlanAreaChanges();
+	}); 
+	
+	afterEach(function() {
+		instance = null;
+	});
+
+	it('has the right rules', function() {
+		const rules = instance.rules;
+        expect(rules.id).to.eql(['required', 'integer']);
```
Comment (TomerMe2) [inline server/tests/unit/model/plan_area_changes.test.js:16]: indentation problem

File: server/api/model/plan_area_changes.js [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+	canRead () {
+		return true;
+	}
+
+	getCollection () {
+		return this.collection();
+	}
+
+    static async byPlanAndUsage (planId, usage) {
+		if (!planId) {
```
Comment (TomerMe2) [inline server/api/model/plan_area_changes.js:35]: indentation problem

File: server/api/model/plan_area_changes.js [hunk 5]
```diff
@@ -0,0 +1,77 @@
+const Model = require('./base_model');
+const Log = require('../lib/log');
+const Exception = require('./exception');
+const { Bookshelf, Knex } = require('../service/database');
+
+class PlanAreaChanges extends Model {
+	get rules () {
+		return {
+			id: ['required', 'integer'],
+            plan_id: ['required', 'integer'],
```
Comment (TomerMe2) [inline server/api/model/plan_area_changes.js:10]: indentation problem

File: server/api/controller/plan_area_changes.js [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+
+class PlanAreaChangesController extends Controller {
+
+    static refreshPlanAreaChanges = async (planId, rawAreaChangesString) => {
+        await PlanAreaChanges.deleteByPlan(planId);
+        try {
+            if (rawAreaChangesString) {
+                let areaChangesArray = JSON.parse(rawAreaChangesString)[0];
+                let areaChanges = [];
+                for (let iChanges=0;iChanges<areaChangesArray.length; iChanges++) {
```
Comment (TomerMe2) [inline server/api/controller/plan_area_changes.js:13]: please add a space after iChanges=0;

**Review by TomerMe2 — CHANGES_REQUESTED** [review]
Thank you for your awesome work. I like the code.
It has some indentation problems, but other than that it's cool.

File: server/bin/populate_plan_area_changes (hunk 1 above)
Comment (muamichali) [reply server/bin/populate_plan_area_changes]: **_done._**

File: server/migrations/20210722090328_create_plan_area_changes_table.js (hunk 2 above)
Comment (muamichali) [reply server/migrations/20210722090328_create_plan_area_changes_table.js]: **_done._**

File: server/tests/unit/model/plan_area_changes.test.js (hunk 3 above)
Comment (muamichali) [reply server/tests/unit/model/plan_area_changes.test.js:16]: **_done._**

File: server/api/model/plan_area_changes.js (hunk 4 above)
Comment (muamichali) [reply server/api/model/plan_area_changes.js:35]: **_done._**

File: server/api/model/plan_area_changes.js (hunk 5 above)
Comment (muamichali) [reply server/api/model/plan_area_changes.js:10]: **_done._**

File: server/api/controller/plan_area_changes.js (hunk 6 above)
Comment (muamichali) [reply server/api/controller/plan_area_changes.js]: **_done._**

File: server/api/model/plan_area_changes.js [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+			.then(() => true);
+	}    
+
+	// adds multiple changes in one transcation
+	static async createPlanAreaChanges( data) {
+		try {
+			await Bookshelf.transaction(async (transaction) => {
+				for (const datum in data ){
+                    const currentChange = data[datum];
+                     await Knex.raw(`INSERT INTO plan_area_changes (plan_id, \`usage\`, measurement_unit, approved_state, change_to_approved_state, total_in_detailed_plan, total_in_mitaarit_plan, remarks)
```
Comment (GalGend) [inline server/api/model/plan_area_changes.js:60]: why not to create  a new plan_area_changes model, set the property and then save? Much easier than entering values to an SQL query

File: server/bin/populate_plan_area_changes.js [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+const { Knex } = require('../api/service/database');
+
+
+
+
+
+const populatePlanAreaChanges = async () => {
+	try {
+		const knexRes = await Knex.raw(`SELECT id, areaChanges
+		FROM plan
```
Comment (GalGend) [inline server/bin/populate_plan_area_changes.js:12]: Use Knex.where

File: server/migrations/20210722090328_create_plan_area_changes_table.js [hunk 9]
```diff
@@ -0,0 +1,17 @@
+exports.up = function(knex) {
+    return knex.schema.createTableIfNotExists('plan_area_changes', t => {
+        t.increments('id').primary();
+        t.integer('plan_id');
+        t.text('usage');
+        t.text('measurement_unit').notNullable();
+        t.text('approved_state').notNullable();
```
Comment (GalGend) [inline server/migrations/20210722090328_create_plan_area_changes_table.js:7]: Leave them nullable, data will not be complete

File: server/api/model/plan_area_changes.js (hunk 7 above)
Comment (muamichali) [reply server/api/model/plan_area_changes.js:60]: I did it at first but I got an error from checkit that it doesn't know how to check text or something like that..

Comment (GalGend) [reply server/api/model/plan_area_changes.js:60]: Can you please show me an example of the error you got? Maybe it's related to the rules on the different text fields?

File: server/migrations/20210722090328_create_plan_area_changes_table.js (hunk 9 above)
Comment (TomerMe2) [reply server/migrations/20210722090328_create_plan_area_changes_table.js:7]: Right now, it is complete for every plan that we got.

**Review by GalGend — APPROVED** [review]

**Comment by GalGend** [comment]
Hi @muamichali @TomerMe2 ,

Merging this branch.
Please make sure to make this process (of add plan area changes) as part of the plan fetch procedure. Also, check this script on large amount of plans (as we have in production ) to verify it will run smoothly in our servers. Consider making the script run as part of the migration, so it runs smoothly to all developers.

**Pull request merged**